pkg/controller/addon/addon_controller.go

```diff
@@ -0,0 +1,223 @@
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileAddon implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileAddon{}
+
+// ReconcileAddon reconciles a Addon object
+type ReconcileAddon struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+	addons map[string]mf.Manifest
+}
+
+// Reconcile reads that state of the cluster for a Addon object and makes changes based on the state read
+// and what is in the Addon.Spec
+// TODO(user): Modify this Reconcile function to implement your Controller logic.  This example creates
+// a Pod as an example
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileAddon) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	log := ctrlLog.WithName("add")
+	reqLogger := log.WithValues("Request.Namespace", req.Namespace, "Request.Name", req.Name)
+	reqLogger.Info("Reconciling Addon")
+
+	// Fetch the Addon instance
+	instance := &op.Addon{}
+	err := r.client.Get(context.TODO(), req.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	if isAddOnUpToDate(instance) {
+		log.Info("skipping installation, resource already up to date")
+		return reconcile.Result{}, nil
+	}
+
+	return r.reconcileAddon(req, instance)
+
+}
+
+func isAddOnUpToDate(r *op.Addon) bool {
+	c := r.Status.Conditions
+	if len(c) == 0 {
+		return false
+	}
+	latest := c[0]
+	return latest.Version == r.Spec.Version &&
+		latest.Code == op.InstalledStatus
+}
+
+func (r *ReconcileAddon) reconcileAddon(req reconcile.Request, res *op.Addon) (reconcile.Result, error) {
+	log := requestLogger(req, "addon")
+
+	err := r.updateStatus(res, op.AddonCondition{Code: op.InstallingStatus, Version: res.Spec.Version})
+	if err != nil {
+		log.Error(err, "failed to set status")
+		return reconcile.Result{}, err
+	}
+
+	addonPath := filepath.Join("deploy", "resources", "addons", res.Name, res.Spec.Version)
+	manifest, err := mf.NewManifest(addonPath, true, r.client)
+	if err != nil {
+		// handle error
+	}
+
+	// TODO: targetNamspace should be decided from current tekton pipelines installation
+	// if there is no pipeline installation then the default namespace 'tekton-pipelines'
+	// can be used
+	targetNamespace := "tekton-pipelines"
+	tfs := []mf.Transformer{
+		mf.InjectOwner(res),
+		mf.InjectNamespace(targetNamespace),
+	}
+
```

Comment:
The owner of the addon is the instance of the addon CRD.  How is the addon deleted when the pipeline is deleted?

Reply:
i have updated the reconcile logic today. I am setting the pipeline installation as owner
the code is functional, but needs some refactoring.